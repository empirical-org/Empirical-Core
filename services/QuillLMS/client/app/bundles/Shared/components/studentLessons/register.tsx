import * as React from 'react';
class Register extends React.Component<any, any> {
  constructor(props) {
    super(props)

    this.state = {
      showIntro: false,
      hasSentenceFragment: this.hasSentenceFragment(),
    }
  }

  componentDidMount() {
    const { previewMode } = this.props;
    if(previewMode) {
      this.startActivity();
    }
  }

  landingPageHtmlHasText(){
    const { lesson, } = this.props
    // strips out everything nested betwee < and >.
    // we should also not allow draft js to send text-less answers from
    // the lp editor
    return lesson.landingPageHtml.replace(/(<([^>]+)>)/ig, "").length > 0;
  }

  startActivity = () => {
    const { lesson, startActivity, } = this.props
    if (lesson.landingPageHtml && this.landingPageHtmlHasText()) {
      this.setState({ showIntro: true, });
    } else {
      startActivity();
    }
  }

  handleStartLessonClick = () => {
    const { startActivity, } = this.props
    startActivity();
  }

  resume = () => {
    const { resumeActivity, session, } = this.props
    resumeActivity(session);
  }

  hasSentenceFragment = () => {
    const { lesson, } = this.props
    const { questions } = lesson;
    for (let i = 0; i < questions.length; i+=1) {
      if (questions[i].questionType === "sentenceFragments") {
        return true;
      }
    }
    return false;
  }

  renderButton = () => {
    const { session, } = this.props
    let onClickFn,
      text;
    if (session) {
      // resume session if one is passed
      onClickFn = this.resume;
      text = <span>Resume</span>;
    } else {
      // otherwise begin new session
      onClickFn = this.startActivity;
      text = <span>Begin</span>;
    }
    return (
      <button className="quill-button-archived focus-on-light primary contained large" onClick={onClickFn} type="button">
        {text}
      </button>
    );
  }

  translatedText = (): string => {
    const { lesson, language} = this.props;
    const { translations, } = lesson;
    return translations && translations[language];
  }


  renderIntro = () => {
    const { lesson, } = this.props
    const { showIntro, hasSentenceFragment, } = this.state
    const translatedText = this.translatedText()
    if (showIntro) {
      return (
        <div className="container">
          <div className="landing-page-html" dangerouslySetInnerHTML={{ __html: lesson.landingPageHtml, }} />
          {translatedText && (
            <React.Fragment>
              <hr />
              <div className="landing-page-html" dangerouslySetInnerHTML={{ __html: translatedText }} />
            </React.Fragment>
          )}
          <button className="quill-button-archived focus-on-light large primary contained" onClick={this.handleStartLessonClick} type="button">Start activity</button>
        </div>
      );
    } else if (hasSentenceFragment) {
      return (
        <div className="container">
          <h2 className="title is-3 register">
            Welcome to Quill Connect Fragments!
          </h2>
          <div className="register-container">
            <ul className="register-list">
              <li>Add to the group of words to make a complete sentence.</li>
              <li>Add the number of words shown in the directions.</li>
              <li>There is often more than one correct answer.</li>
              <li>Remember to use correct spelling, capitalization, and punctuation!</li>
            </ul>
            {this.renderButton()}
            <br />
          </div>
        </div>
      );
    } else {
      return (
        <div className="container">
          <h2 className="title is-3 register">
            Welcome to Quill Connect!
          </h2>
          <div className="register-container">
            <ul className="register-list">
              <li>Combine the sentences together into one sentence.</li>
              <li>You may add or remove words.</li>
              <li>There is often more than one correct answer.</li>
              <li>Remember to use correct spelling, capitalization, and punctuation!</li>
            </ul>
            {this.renderButton()}
            <br />
          </div>
        </div>
      );
    }
  }

  render() {
    return (
      <section
        className="student"
        style={{
          paddingTop: 20,
        }}
      >
        {this.renderIntro()}
      </section>
    );
  }
}

export { Register };
