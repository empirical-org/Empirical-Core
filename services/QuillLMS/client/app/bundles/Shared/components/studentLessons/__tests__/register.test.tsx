import * as React from "react";
import { render, screen } from "@testing-library/react";

import { Register } from "../register";

const mockProps = {
  lesson: {
    questions: [],
    questionType: '',
    landingPageHtml: null
  },
  startActivity: jest.fn(),
  resumeActivity: jest.fn(),
  session: {},
  translate: jest.fn(),
  showTranslation: false
}

describe('Register', () => {
  test('it should render default Connect intro', () => {
    const { asFragment } = render(<Register {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
    const heading = screen.getByRole('heading', {
      name: /welcome to quill connect!/i
    });
    expect(heading).toBeInTheDocument()
  })
  test('it should render lesson specific intro', () => {
    mockProps.lesson.questionType = 'fillInBlank'
    mockProps.lesson.landingPageHtml = 'test landing page'
    const { asFragment } = render(<Register {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
    const landingPageText = screen.getByText(/test landing page/i);
    expect(landingPageText).toBeInTheDocument()
  })
})
