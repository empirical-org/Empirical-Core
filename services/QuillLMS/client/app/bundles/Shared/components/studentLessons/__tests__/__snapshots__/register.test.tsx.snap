// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register it should render default Connect intro 1`] = `
<DocumentFragment>
  <section
    class="student"
    style="padding-top: 20px;"
  >
    <div
      class="container"
    >
      <h2
        class="title is-3 register"
      >
        Welcome to Quill Connect!
      </h2>
      <div
        class="register-container"
      >
        <ul
          class="register-list"
        >
          <li>
            Combine the sentences together into one sentence.
          </li>
          <li>
            You may add or remove words.
          </li>
          <li>
            There is often more than one correct answer.
          </li>
          <li>
            Remember to use correct spelling, capitalization, and punctuation!
          </li>
        </ul>
        <button
          class="quill-button-archived focus-on-light primary contained large"
          type="button"
        >
          Resume
        </button>
        <br />
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Register it should render lesson specific intro 1`] = `
<DocumentFragment>
  <section
    class="student"
    style="padding-top: 20px;"
  >
    <div
      class="container"
    >
      <div
        class="landing-page-html"
      >
        test landing page
      </div>
      <button
        class="quill-button-archived focus-on-light primary contained large"
        type="button"
      >
        Resume
      </button>
    </div>
  </section>
</DocumentFragment>
`;
